botocore/operation.py

```diff
@@ -86,8 +127,28 @@ def call(self, endpoint, **kwargs):
         self.session.emit(event, endpoint=endpoint,
                           model=self.model,
                           params=request_dict,
-                          operation=self)
-        response = endpoint.make_request(self.model, request_dict)
+                          operation=self,
+                          request_signer=signer)
+
+        # Here we register to the specific request-created event
+        # for this operation. Since it's possible to run the same
+        # operation in multiple threads, we used a lock to prevent
+        # issues. It's possible a request will be signed more than
+        # once. Once the request has been made, we unregister the
+        # handler.
+        def request_created(request, **kwargs):
+            with threading.Lock():
```

Comment:
Are we absolutely sure there isn't a problem with mutating the request as it's in flight through the requests/httplib code?  Seems like this is very sensitive to race conditions.  Also we read through the request body as we sign the code.  I assume this could cause issues if we're also trying to send the request at the same time.

This also means that the number of times this request is proportional to the concurrency level which I'm not really a big fan of.  I assume for things like sigv4 payload checksums this can get expensive.